File: programmingChallenge/treapFarm.py
import sys, random
sys.setrecursionlimit(1_000_000)
rnd = random.Random(42) # fixed seed for per‑crop balancing

# --------- Per‑crop Treap for(COUNT) ----------
class PNode:
    __slots__ = ('key','prio','l','r','sz')
    def __init__(self, key):
        self.key = key
        self.prio = rnd.randint(0, 2**31-1)
        self.l = self.r = None
        self.sz = 1

def psz(t): return 0 if t is None else t.sz
def ppull(t): 
    if t: t.sz = 1 + psz(t.l) + psz(t.r)

def psplit(t, key): # ≤key | >key
    if t is None: return (None, None)
    if key < t.key:
        a,b = psplit(t.l, key)
        t.l = b; ppull(t)
        return (a, t)
    a,b = psplit(t.r, key)
    t.r = a; ppull(t)
    return (t, b)

def pmerge(a, b):
    if a is None: return b
    if b is None: return a
    if a.prio > b.prio:
        a.r = pmerge(a.r, b); ppull(a); return a
    b.l = pmerge(a, b.l); ppull(b); return b

def pinsert(root, key):
    node = PNode(key)
    l,r = psplit(root, key)
    return pmerge(pmerge(l, node), r)

def perase(root, key):
    if root is None: return None
    if root.key == key:
        return pmerge(root.l, root.r)
    if key < root.key:
        root.l = perase(root.l, key)
    else:
        root.r = perase(root.r, key)
    ppull(root); return root

def pcount_less(t, key):
    if t is None: return 0
    if key <= t.key: return pcount_less(t.l, key)
    return psz(t.l) + 1 + pcount_less(t.r, key)


# --------- Global Treap (STRONGEST) --------
class GNode:
    __slots__ = ('key','crop','time','l','r','sz','maxTime','maxIdx','maxCrop')
    def __init__(self, key, crop, time):
        self.key, self.crop, self.time = key, crop, time
        self.l = self.r = None
        self.sz = 1
        self.maxTime, self.maxIdx, self.maxCrop = time, key, crop

def gsz(t): return 0 if t is None else t.sz
def gpull(t):
    if not t: return
    t.sz = 1 + gsz(t.l) + gsz(t.r)
    t.maxTime, t.maxIdx, t.maxCrop = t.time, t.key, t.crop
    for ch in (t.l, t.r):
        if ch and (ch.maxTime > t.maxTime or (ch.maxTime==t.maxTime and ch.maxIdx < t.maxIdx)):
            t.maxTime, t.maxIdx, t.maxCrop = ch.maxTime, ch.maxIdx, ch.maxCrop

def gsplit(t, key): # ≤key | >key
    if t is None: return (None, None)
    if key < t.key:
        a,b = gsplit(t.l, key)
        t.l = b; gpull(t); return (a, t)
    a,b = gsplit(t.r, key)
    t.r = a; gpull(t); return (t, b)

def gmerge(a, b): # heap by (time, -idx)
    if a is None: return b
    if b is None: return a
    if (a.time, -a.key) > (b.time, -b.key):
        a.r = gmerge(a.r, b); gpull(a); return a
    b.l = gmerge(a, b.l); gpull(b); return b

def ginsert(root, idx, crop, time):
    node = GNode(idx, crop, time)
    l,r = gsplit(root, idx)
    return gmerge(gmerge(l, node), r)

def gerase(root, idx):
    if root is None: return None
    if root.key == idx:
        return gmerge(root.l, root.r)
    if idx < root.key:
        root.l = gerase(root.l, idx)
    else:
        root.r = gerase(root.r, idx)
    gpull(root); return root


# --------- Game -------------
N, Q = map(int, sys.stdin.readline().split())
cropType = list(map(int, sys.stdin.readline().split()))
cropTime = [0]*N
globalTime = 0
perCrop = {}
globalRoot = None
for i,v in enumerate(cropType):
    perCrop[v] = pinsert(perCrop.get(v), i)
    globalRoot = ginsert(globalRoot, i, v, 0)

undo = [] # stack of (idx, crop, time), for the undo operation
out = [] # output buffer

def apply_replace(idx, newCrop):
    global globalTime, globalRoot
    undo.append((idx, cropType[idx], cropTime[idx]))
    oldCrop = cropType[idx]
    perCrop[oldCrop] = perase(perCrop[oldCrop], idx)
    perCrop[newCrop] = pinsert(perCrop.get(newCrop), idx)
    globalRoot = gerase(globalRoot, idx)
    globalTime += 1
    globalRoot = ginsert(globalRoot, idx, newCrop, globalTime)
    cropType[idx] = newCrop; cropTime[idx] = globalTime

def undo_replace(idx, oldCrop, oldTime):
    global globalRoot
    perCrop[cropType[idx]] = perase(perCrop[cropType[idx]], idx)
    globalRoot = gerase(globalRoot, idx)
    perCrop[oldCrop] = pinsert(perCrop.get(oldCrop), idx)
    globalRoot = ginsert(globalRoot, idx, oldCrop, oldTime)
    cropType[idx] = oldCrop; cropTime[idx] = oldTime

for _ in range(Q):
    parts = sys.stdin.readline().split()
    op = parts[0]
    if op == 'C':
        i,v = map(int, parts[1:])
        ans = pcount_less(perCrop.get(v), i) if v in perCrop else 0
        out.append(str(ans))
    elif op == 'R':
        i,v = map(int, parts[1:])
        apply_replace(i, v)
    elif op == 'U':
        k = int(parts[1])
        for _ in range(k):
            idx, oc, ot = undo.pop()
            undo_replace(idx, oc, ot)
    else:                               # S l r
        l,r = map(int, parts[1:])
        a,b = gsplit(globalRoot, l-1)
        mid,c = gsplit(b, r)
        out.append(str(mid.maxCrop))
        globalRoot = gmerge(a, gmerge(mid, c))

sys.stdout.write("\n".join(out))

File: programmingChallenge/treapFarm.java
import java.io.*;
import java.util.*;

public class treapFarm {
    // ----------------- Per‑crop Treap (for COUNT) --------------------
    static final Random RNG = new Random(42); // fixed seed for per‑crop balancing
    static class PNode {
        int key, prio = RNG.nextInt(); // key and priority value for the Treap
        PNode left, right;
        int size = 1;
    }
    static int getSize(PNode t){ return t == null ? 0 : t.size; }
    static void pull(PNode t){ if(t!=null) t.size = 1 + getSize(t.left) + getSize(t.right); }
    static PNode[] pSplit(PNode t, int key){
        if(t==null) return new PNode[]{null,null};
        if(key < t.key){
            PNode[] sp = pSplit(t.left, key);
            t.left = sp[1]; pull(t);
            return new PNode[]{sp[0], t};
        } else {
            PNode[] sp = pSplit(t.right, key);
            t.right = sp[0]; pull(t);
            return new PNode[]{t, sp[1]};
        }
    }
    static PNode pMerge(PNode a, PNode b){
        if(a==null) return b;
        if(b==null) return a;
        if(a.prio > b.prio){
            a.right = pMerge(a.right, b); pull(a); return a;
        } else {
            b.left = pMerge(a, b.left); pull(b); return b;
        }
    }
    static PNode pInsert(PNode root, int key){
        PNode node = new PNode(); node.key = key;
        PNode[] sp = pSplit(root, key);
        return pMerge(pMerge(sp[0], node), sp[1]);
    }
    static PNode pErase(PNode root, int key){
        if(root==null) return null;
        if(root.key == key){
            return pMerge(root.left, root.right);
        } else if(key < root.key){
            root.left = pErase(root.left, key);
        } else {
            root.right = pErase(root.right, key);
        }
        pull(root);
        return root;
    }
    static int pCountLess(PNode t, int key){
        if(t==null) return 0;
        if(key <= t.key) return pCountLess(t.left, key);
        return getSize(t.left) + 1 + pCountLess(t.right, key);
    }

    // ------------------ Global Treap (for STRONGEST) ------------
    static class GNode {
        int key; // plot index
        int crop; // crop type
        int time; // last update timestamp
        GNode left, right;
        int size = 1;
        int maxTime;
        int maxIdx; // leftmost index with maxTime
        int maxCrop; // crop at maxIdx
    }
    static int getSize(GNode t){ return t==null ? 0 : t.size; }
    static void pull(GNode t){
        if(t==null) return;
        t.size = 1 + getSize(t.left) + getSize(t.right);

        // start with self
        t.maxTime = t.time;
        t.maxIdx  = t.key;
        t.maxCrop = t.crop;

        // left subtree
        if(t.left != null){
            if(t.left.maxTime > t.maxTime ||
                    (t.left.maxTime == t.maxTime && t.left.maxIdx < t.maxIdx)){
                t.maxTime = t.left.maxTime;
                t.maxIdx  = t.left.maxIdx;
                t.maxCrop = t.left.maxCrop;
            }
        }
        // right subtree
        if(t.right != null){
            if(t.right.maxTime > t.maxTime ||
                    (t.right.maxTime == t.maxTime && t.right.maxIdx < t.maxIdx)){
                t.maxTime = t.right.maxTime;
                t.maxIdx  = t.right.maxIdx;
                t.maxCrop = t.right.maxCrop;
            }
        }
    }
    static GNode[] gSplit(GNode t, int key){
        if(t==null) return new GNode[]{null,null};
        if(key < t.key){
            GNode[] sp = gSplit(t.left, key);
            t.left = sp[1]; pull(t);
            return new GNode[]{sp[0], t};
        } else {
            GNode[] sp = gSplit(t.right, key);
            t.right = sp[0]; pull(t);
            return new GNode[]{t, sp[1]};
        }
    }
    static GNode gMerge(GNode a, GNode b){
        if(a==null) return b;
        if(b==null) return a;
        if(a.time > b.time){
            a.right = gMerge(a.right, b); pull(a); return a;
        } else if(b.time > a.time){
            b.left = gMerge(a, b.left); pull(b); return b;
        } else {
            // if tie by timestamp, then break by index (smaller index as root to keep leftmost)
            if(a.key < b.key){
                a.right = gMerge(a.right, b); pull(a); return a;
            } else {
                b.left = gMerge(a, b.left); pull(b); return b;
            }
        }
    }
    static GNode gInsert(GNode root, int idx, int crop, int time){
        GNode node = new GNode();
        node.key = idx; node.crop = crop; node.time = time;
        node.maxTime = time; node.maxIdx = idx; node.maxCrop = crop;
        GNode[] sp = gSplit(root, idx);
        return gMerge(gMerge(sp[0], node), sp[1]);
    }
    static GNode gErase(GNode root, int idx){
        if(root==null) return null;
        if(root.key == idx){
            return gMerge(root.left, root.right);
        } else if(idx < root.key){
            root.left = gErase(root.left, idx);
        } else {
            root.right = gErase(root.right, idx);
        }
        pull(root);
        return root;
    }

    // -------------------- Game vars -------------------------
    static int N, Q;
    static int[] cropType; // current crop at each plot
    static int[] cropTime; // current timestamp at each plot
    static int globalTime = 0;

    // map crop → its per‑crop Treap root
    static HashMap<Integer,PNode> perCrop = new HashMap<>();
    static GNode globalRoot = null;

    // undo stack
    static class Action {
        int idx, oldCrop, oldTime;
        Action(int i, int c, int t){ idx=i; oldCrop=c; oldTime=t; }
    }
    static Deque<Action> undoStack = new ArrayDeque<>();

    public static void main(String[] args) throws IOException {
        Scanner fs = new Scanner(System.in);
        N = fs.nextInt(); Q = fs.nextInt();
        cropType = new int[N];
        cropTime = new int[N];

        // read initial crops (timestamp = 0)
        for(int i=0; i<N; i++){
            int v = fs.nextInt();
            cropType[i] = v;
            cropTime[i] = 0;
            // per‑crop insertion
            perCrop.putIfAbsent(v, null);
            perCrop.put(v, pInsert(perCrop.get(v), i));
            // global insertion
            globalRoot = gInsert(globalRoot, i, v, 0);
        }

        StringBuilder out = new StringBuilder();
        while(Q-- > 0){
            String op = fs.next();
            if(op.equals("C")){
                int i = fs.nextInt(), v = fs.nextInt();
                PNode t = perCrop.get(v);
                int ans = (t==null ? 0 : pCountLess(t, i));
                out.append(ans).append('\n');

            } else if(op.equals("R")){
                int i = fs.nextInt(), v = fs.nextInt();
                applyReplace(i, v);

            } else if(op.equals("U")){
                int k = fs.nextInt();
                while(k-- > 0){
                    Action a = undoStack.pop();
                    // revert to a.oldCrop, a.oldTime
                    undoReplace(a.idx, a.oldCrop, a.oldTime);
                }

            } else if(op.equals("S")){
                int l = fs.nextInt(), r = fs.nextInt();
                // split into [0..l-1], [l..r], [r+1..N-1]
                GNode[] sp1 = gSplit(globalRoot, l-1);
                GNode[] sp2 = gSplit(sp1[1], r);
                int strongestCrop = sp2[0].maxCrop;
                out.append(strongestCrop).append('\n');
                globalRoot = gMerge(sp1[0], gMerge(sp2[0], sp2[1]));
            }
        }
        System.out.print(out);
    }

    // perform replace with timestamp++ and record undo
    static void applyReplace(int idx, int newCrop){
        // record old state
        undoStack.push(new Action(idx, cropType[idx], cropTime[idx]));
        // remove old from per‑crop
        int oldCrop = cropType[idx];
        perCrop.put(oldCrop, pErase(perCrop.get(oldCrop), idx));
        // insert new into per‑crop
        perCrop.putIfAbsent(newCrop, null);
        perCrop.put(newCrop, pInsert(perCrop.get(newCrop), idx));
        // update global
        globalRoot = gErase(globalRoot, idx);
        int t = ++globalTime;
        globalRoot = gInsert(globalRoot, idx, newCrop, t);
        // update arrays
        cropType[idx] = newCrop;
        cropTime[idx] = t;
    }

    // undo a replacement to oldCrop and oldTime
    static void undoReplace(int idx, int oldCrop, int oldTime){
        // remove current
        perCrop.put(cropType[idx], pErase(perCrop.get(cropType[idx]), idx));
        globalRoot = gErase(globalRoot, idx);
        // restore per‑crop
        perCrop.putIfAbsent(oldCrop, null);
        perCrop.put(oldCrop, pInsert(perCrop.get(oldCrop), idx));
        // restore global
        globalRoot = gInsert(globalRoot, idx, oldCrop, oldTime);
        // update arrays
        cropType[idx] = oldCrop;
        cropTime[idx] = oldTime;
    }
}
